api/server/sdk/status/errormessage/errormessage.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2019 Portworx
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package errormessage
+
+import (
+	"fmt"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/libopenstorage/openstorage/api"
+	sdkstatus "github.com/libopenstorage/openstorage/api/server/sdk/status"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// Error maintains the initial gRPC error while appending new errors
+// to the error details as it goes up the stack. This enables the gRPC to maintain
+// the initial error code and condition and separating them from the errors in the
+// stack as the server returns from the code.
+func Error(err error, c codes.Code, msg string) error {
+
+	// No error was given, we just return a new status
+	if err == nil {
+		return status.Error(c, msg)
+	}
+
+	// Check if a non-gRPC status was provided
+	orig, ok := status.FromError(err)
+	if !ok {
+		// not a grpc style error so use the gRPC error information as the
+		// actual error and add err to the details
+		return ErrorSet(err, c, msg)
+	}
+
+	// Add the current gRPC error to the original
+	var derr error
+	orig, derr = orig.WithDetails(&api.SdkErrorMessage{
+		Status: status.New(c, msg).Proto(),
+	})
+	if derr != nil {
+		panic("Unexpected")
+	}
+	return orig.Err()
+}
+
+// Errorf calls Error() with a formatted string
+func Errorf(err error, c codes.Code, format string, a ...interface{}) error {
+	return Error(err, c, fmt.Sprintf(format, a...))
+}
+
+// ErrorSet places the current gRPC status as the actual error
+// pushes all the values in err as details.
+func ErrorSet(err error, c codes.Code, msg string) error {
+
+	if err == nil {
+		return status.Error(c, msg)
+	}
+
+	// Get gRPC status from error. This function also creates a new gRPC status
+	// if it is a non-gRPC status error.
+	orig := sdkstatus.FromError(err)
+
+	// Create a new error
+	s := status.New(c, msg)
+
+	// Add err as the first error to maintain sequence
+	origDetails := orig.Details()
```

Comment:
Isn't ErrorSet going to get called when the `err` passed in is not a grpc error. So orig.Details() will be empty.
So we won't need the loop over origDetails and we can simply add a single SdkErrorMessage to the details array.

 Also after calling ErrorSet for the first time, then subsequent errors will be grpc errors as we converted the original error into a grpc error by calling sdkstatus.FromError().

May be we make ErrorSet a private function, and let the users of the library use only Error and Errorf

Is my understanding correct ?